Update the KEP for validation (#2)
jpbetz/enhancements: keps/sig-api-machinery/NNNN-crd-validation-defaulting-conversion-expressions/README.md

## Summary

> CRDs need direct support for non-trivial validation.  Comparing to the current mechanism of using
> admission webhooks to handle the CRDs validation, they significantly complicate the development and

Reviewer: ```suggestion
CRDs need direct support for non-trivial validation.  While admission webhooks do support
CRDs validation, they significantly complicate the development and
```

## Proposal

[Common Expression Language (CEL)](https://github.com/google/cel-go) is an

> excellent supplement to the current validation mechanism because it is sufficiently expressive to

Reviewer 2: Just trying to understand more, what's the advantage of CEL over other languages for defining unified policy?
For example why not we use OPA by CNCF?

Author: OPA would be used as admission controller which shares very similar pro/cons as webhooks.

Reviewer: There are two orthogonal aspects to this:

1. The idea of using a general purpose webhook for validation

  - pro: Doesn't require we make a change to Kubernetes core
  - con: webhooks are only called for types that they are registered for, so each CRD needs to be registered
  - con: requires introducing the webhook in the first place (and getting everyone to install it and deal with upgrades/roll-backs, ...)
  - con: doesn't promote a unified ecosystem for validation like 1st class support does

2. The OPA rule grammar vs. CEL

- designed for multiline programs (both a con and a pro)
- does not have the same sandboxing constraints as CEL (clearly a con)
- doesn't typecheck (con)
- has a full package and module system (a con for when writing readable and compose-able validation rules)
- integrates tightly with Kubernetes types (a significant pro)
- opinionated about the language being for policy declarations, this bleeds into the grammar quite a bit (con, but could be overcome by working with OPA community to generalize grammar)

We really a section in the KEP that captures the details of these types of comparisons.  I also think we need to be clear that we're still in the early phases of experimenting with all the various available grammars and that CEL is currently our leading candidate but that it is very possible that we will switch to something else if we find an alternative that is objectively better.

      properties:
        spec:
          x-kubernetes-validator: 

> 		    - rule: this.properties.minReplicas <= this.properties.maxReplicas

Reviewer 2: this is an expression that has direct access to all the fields of this object (in spec). do we really need to scope to the fields using 'this'?

Author: It has been updated. We could talk over it later

> 		    - rule: this.properties.minReplicas <= this.properties.maxReplicas

Reviewer: Let's keep this as minimal as possible for now.  We might need a `this` keyword of equivalent, but I don't that's decided. And adding `properties` adds a lot of verbosity for unclear gains.

Consider including folks who also work outside the SIG or subproject.
-->

### Future Plan

#### Defaulting

The `x-kubernetes-default` extension will be used. The location of the defaulter
determines the scope of defaulter.

The 'field' specifies which field is to be defaulted using a field path that is
relative to the scope root. The defaulter will only be run if 'field' is unset, 
and the result of the expression must match the type of the field to be
defaulted.

```yaml
apiVersion: apiextensions.k8s.io/v1
kind: CustomResourceDefinition
...
  schema:
    openAPIV3Schema:
      type: object
      properties:
        spec:
          x-kubernetes-default:
		    - rule: "!has(spec.type) && spec.Type == 'LoadBalancer' ? 'ServiceExternalTrafficPolicy' : null"
			  field: spec.externalTrafficPolicy
```

- Each defaulter may have multiple rules.

- The 'field' is optional and defaults to the scope if not explicitly set.

If the expression evaluates to null, the field is left unset.

> TODO: Any downsides to using null this way?

Reviewer 2: leaving the field unset makes our work extra hard to get the kind of data if we need to. Currently I can not think of any usecases where we need this. From a related angle, leaving the fields unset to represent null value reduce our convictions about the data.

Reviewer: I think the idea here was to make it possible to conditionally default some field value only is some condition is true, otherwise don't do any defaulting. But the specifics of how we do that is a bit of an incomplete thought here.

Let's drop this from the KEP for now since this has been moved to future work and is an incomplete thought. We can circle back when we get to defaulting.

Reviewer: I'll clean this up in a separate PR actually, since this PR just moved the text.
